openwisp_monitoring/device/tests/test_api.py

```diff
@@ -364,19 +358,19 @@ def test_device_metrics_received_signal(self):
 
     def test_invalid_chart_config(self):
         # Tests if chart_config is invalid, then it is skipped and not failed
-        d = self._create_device(organization=self._create_org())
-        m = self._create_object_metric(name='test_metric', content_object=d)
-        c = self._create_chart(metric=m, test_data=None)
+        device = self._create_device(organization=self._create_org())
+        metric = self._create_object_metric(name='test_metric', content_object=device)
+        conf = self._create_chart(metric=metric, test_data=None)
         with redirect_stderr(StringIO()) as stderr:
-            c.configuration = 'invalid'
-            c.save()
-            r = self.client.get(self._url(d.pk.hex, d.key))
-        self.assertIn('InvalidChartConfigException', stderr.getvalue())
-        self.assertEqual(r.status_code, 200)
+            conf.configuration = 'invalid'
+            conf.save()
+            # TODO:N not working testcase
+            self.assertIn('InvalidChartConfigException', stderr.getvalue())
```

Comment:
@nepython this test is failing, I checked couple of things and I doesn't seem to failing because of openwisp-controller update.
Infact, when I run this manually, typing `conf.configuration` does throw the error, but `stderr.getvalue()` is still empty, is this failing because of some other work I should be aware of or should I proceed to modify the test to fix it?

Reply:
You need to send the request prior to checking the exception, the reason being the exception is raised after `clean` method is called. So, you can simply do it as is being done currently.
```python
response = self.client.get(self._url(device.pk.hex, device.key))
self.assertIn('InvalidChartConfigException', stderr.getvalue())
```

Also, note :point_up: above you have used `device` but there is no such variable defined (it's still `d`, so change either as seems better :smiley:)

Reply:
Changing the order fixed it. I may have mistakenly changed the order while fixing another testcase, thanks! :smile: